### PR TITLE
feat: retry GitHub API calls with exponential backoff

### PR DESCRIPTION
**Pull request by pvdlg**
Fix semantic-release/semantic-release#607, Fix semantic-release/semantic-release#658

Wrap the `@octokit/rest` functions used by the plugin with [p-retry](https://github.com/sindresorhus/p-retry) with the default settings:
- Retries 3 times (in addition of the first attempt) on all errors except `400`, `401` and `403`
- Min timeout of 1s
- Factor of 2 

So the calls to the same function will be done on the following timeline: 0s, 1s, 3s and 7s. If it still fails on the 4th attempt then we throw the error return by `@octokit/rest`.

### CODE REVIEW

File: lib/get-client.js [hunk 1]
```diff
@@ -10,5 +32,26 @@ module.exports = (githubToken, githubUrl, githubApiPathPrefix) => {
...
+          if (SKIP_RETRY_CODES.includes(err.code)) {
+            // If the API return a 400, 401 or 403 error, do not retry
+            throw new pRetry.AbortError(err);
+          }
+          throw err;
+        }
+      }, retryConf)
+    );
+  }
+}
```
Comment (gr2m) [inline lib/get-client.js]: I think this implementation is rather hard to read ... maybe see the implementation from Greenkeeper for comparison: https://github.com/greenkeeperio/greenkeeper/blob/1704eaca48389699725656242b5c2e1f9213e3b2/lib/github.js#L6-L27 (not that the `.hook` API is not public API yet so we'd need to pin the `@octokit/rest` package).

Do we need to hard-code the methods we want to implement retry? Maybe we could do retries on all writing operations + search? 

The next milestone for `@ocotkit/rest` will be all things APIs and I will look into making best practises like throttling and retries  part of `@ocotkit/rest` itself. So if you like we can merge this as is but would be good to review it once throttling is implemented in `@octokit/rest` itself

Comment (pvdlg) [reply lib/get-client.js]: Using the  `octokit.hook.error`seems more elegant than replacing the function in the client object like I do. I can change the code to use that. That would avoid having to list the functions we want to wrap.

Do we have to pin the exact version or just `~14.0.0`?

Once it's in `@ocotkit/rest` will review and make the changes necessary to use it.

**Comment by pvdlg** [comment]
I made the change suggested [here](https://github.com/semantic-release/github/pull/37#discussion_r167755882).
The problem with using the error hook to retry is that the first retry is done right away after the first failure. It's is probably not that useful and will likely result in another failure, "wasting" and API call. This is the case in the GreenKeeper implementation, by the way.

So I had to add a bit of "creativity" to make it work as expected and make calls on the 0s, 1s, 3s and 7s mark.

That also forces us to duplicate the `SKIP_RETRY_CODES.includes(err.code)` test, but that's not a big deal though.

**Comment by pvdlg** [comment]
Testing a bit more there is another problem, that seems to be a bug in `@octokit/rest` plugin implementation.

If the hook callback (the second parameter of `octokit.hook.error`) throw and error it's ignored and not thrown by the client. Instead the if the error passed in the first argument of the callback (the `error` in `(error, options) => {}`) that is thrown.

This is demonstrated by [this test](https://github.com/semantic-release/github/pull/37/files#diff-4efd3cc6f2a533ff1ebf58ff6d803667R229).
Here the mock server return an error `404` on the first call and a `400` on the second call.
The expected behavior is to retry after the `404` thrown by the first call and to throw the `400` during the second attempt.
On the second attempt `pRetry` is called, make the second call, catch the `400` and throws it with `throw new pRetry.AbortError(err)`. `pRetry takes care of unwrapping the `AbortError`) so what's caught by the hook callback is the error `400`. But the instead of throwing that is swallows it and throw the original `400` error.

Long story short this technique, in each situation in which the server a retryable error then a non-retryable one or the other way around the program doesn't behave as expected, as we want to display to the client API user the "unexpected error" (aka the the one on which we decided retrying is pointless). I think GreenKeeper has the same problem.

I kept the second commit to demonstrate the problem. But I guess we should use the first version (the first commit) until retries are implemented in `@octokit/rest`.

**Comment by pvdlg** [comment]
Closing this one in favor of #42.
I'll keep the branch in case it's useful for `@octokit/rest.js`
